Fix disabled 'eth_sign' event & add 'Request Disabled' event type

digiwand opened the pull request:
## Explanation

Fixes https://github.com/MetaMask/metamask-extension/issues/17974

- Fixes bug where 'Signature Approved' event was incorrectly called when 'eth_sign' rpc method is called while the method is disabled by advanced settings
- Replaces 'Signature Approved' event with new 'Signature Failed' event when the disabled 'eth_sign' event is called (spec discussed [here in Slack](https://consensys.slack.com/archives/G01FXS64DFA/p1678102135498189))

**note:** 
1. sometimes, the 'Signature Failed' effect may be received before 'Signature Request', but the 'sent_at' time still records it being sent earlier. 
2. clean-up code, including more use of constants, is applied to this follow-up PR https://github.com/MetaMask/metamask-extension/pull/18008

<!--
Thanks for the pull request. Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?

Are there any issues, Slack conversations, Zendesk issues, user stories, etc. reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

## Screenshots/Screencaps

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

### Before


### After

<!-- How does it look now? Drag your file(s) below this line: -->
<img width="1473" alt="Screen Shot 2023-03-13 at 12 10 39 PM" src="https://user-images.githubusercontent.com/20778143/224849236-ea8277f5-ebbb-4dd4-a63c-80210f13e929.png">
<img width="1470" alt="Screen Shot 2023-03-13 at 12 10 45 PM" src="https://user-images.githubusercontent.com/20778143/224849240-1e598021-54ab-4db9-9311-193aa8712644.png">


## Manual Testing Steps

1. Open Segment or open background console for check the network requests to Segment
2. Disable the functionality of eth_sign from MetaMask Advanced Settings
3. On another window, go to test dapp
4. Connect to test dapp
5. Click Eth Sign
6. See Signature Request and Request Denied events are fired. Signature Approval should no longer be fired.


<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [x] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [x] PR is linked to the appropriate GitHub issue
- [x] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

digiwand committed:
Fix #17974 / add REQUEST_DISABLED event

digiwand committed:
RPCMiddleware: add eth_sign REQUEST_DISABLED test

digiwand committed:
Merge branch 'develop' into fix-disabled-eth-sign-event

digiwand commented:
moving back to draft. More discussions are being had to plan how we will want to handle this event

digiwand committed:
don't call Signature Request for disabled eth_sign

digiwand committed:
clean: alphabetize

jpuri commented on app/scripts/lib/createRPCMethodTrackingMiddleware.js:
```diff
@@ -138,6 +140,12 @@ export default function createRPCMethodTrackingMiddleware({
+      method === MESSAGE_TYPE.ETH_SIGN &&
+      res.error?.code === errorCodes.rpc.methodNotFound;
+
+    const isDisabledRequest = isDisabledEthSignAdvancedSetting;
```
hey, reason for duplicating the variable ?

digiwand replied:
heyy, I did this for readability down the line

digiwand replied:
hey, I did this for readability down the line. hope it helps, unless it's just me - In that case, happy to change it back

jpuri reviewed (dismissed):
Great work 👍

digiwand committed:
clean: rn isDisabledRequest -> isDisabledRPCMethod

digiwand committed:
fix test

digiwand committed:
rpc middeware: add eth_sign comment

blackdevelopa reviewed (dismissed):
LGTM

seaona reviewed (changes requested):
oh I am still seeing the `Signature Requested` and `Signature Approved` event sent to Segment, with `eth_sign` disabled and no `Request Disabled` event thrown.

I believe I have the last changes :thinking: : could you double check this behaviour to discard it's just me? :pray: :

https://user-images.githubusercontent.com/54408225/224732586-4e35c8fe-63c5-4400-80ef-66a18b8e515b.mp4

digiwand commented:
oof, great catch @seaona! 

Derp.. when we decided to exclude `Signature Request`, I updated the code and moved the `res.error?.code` check up, except the response, `res`, doesn't receive the values at this point... 

I'm reworking this PR now

digiwand committed:
fix eth_sign events and re-add Signature Request

digiwand committed:
send 'Signature Failed' for disabled eth_sigh
